File: algorithms/solutions/week23/toanha/simplify-path.go
package main

import (
	"strings"
)

func simplifyPath(path string) string {
	var dirs = strings.Split(path, "/")
	var l = len(dirs)
	for i := 0; i < l; i++ {
		if dirs[i] == "." || dirs[i] == "" {
			dirs = append(dirs[:i], dirs[i+1:]...)
			i--
			l--
		} else {
			if dirs[i] == ".." {
				if i > 0 {
					dirs = append(dirs[:i-1], dirs[i+1:]...)
					i -= 2
					l -= 2
				} else {
					dirs = append(dirs[:i], dirs[i+1:]...)
					i--
					l--
				}
			}
		}
	}

	return "/" + strings.Join(dirs, "/")
}

File: algorithms/solutions/week23/unrealhoang/simplify_path.rb
# @param {String} path
# @return {String}
def simplify_path(path)
  path_parts = path.split("/").select { |p| p != "" }
  path_components = []
  path_parts.each do |p|
    if p == "."
      next
    elsif p == ".."
      path_components.pop if path_components.length > 0
    else
      path_components << p
    end
  end
  "/" + path_components.join("/")
end
